R/slide.R

```diff
@@ -363,3 +364,403 @@ epi_slide <- function(x, f, ..., before, after, ref_time_values,
+    ref_time_values <- unique(x$time_value)
+  } else {
+    assert_numeric(ref_time_values, min.len = 1L, null.ok = FALSE, any.missing = FALSE)
+    if (!test_subset(ref_time_values, unique(x$time_value))) {
+      cli_abort(
+        "`ref_time_values` must be a unique subset of the time values in `x`."
+      )
+    }
+    if (anyDuplicated(ref_time_values) != 0L) {
+      cli_abort("`ref_time_values` must not contain any duplicates; use `unique` if appropriate.")
+    }
+  }
+  ref_time_values <- sort(ref_time_values)
+
+  # Validate and pre-process `before`, `after`:
+  if (!missing(before)) {
+    before <- vctrs::vec_cast(before, integer())
+    assert_int(before, lower = 0, null.ok = FALSE, na.ok = FALSE)
+  }
+  if (!missing(after)) {
+    after <- vctrs::vec_cast(after, integer())
+    assert_int(after, lower = 0, null.ok = FALSE, na.ok = FALSE)
+  }
+  if (missing(before)) {
+    if (missing(after)) {
+      cli_abort("Either or both of `before`, `after` must be provided.")
+    } else if (after == 0L) {
+      cli_warn("`before` missing, `after==0`; maybe this was intended to be some
+            non-zero-width trailing window, but since `before` appears to be
+            missing, it's interpreted as a zero-width window (`before=0,
+            after=0`).")
+    }
+    before <- 0L
+  } else if (missing(after)) {
+    if (before == 0L) {
+      cli_warn("`before==0`, `after` missing; maybe this was intended to be some
+            non-zero-width leading window, but since `after` appears to be
+            missing, it's interpreted as a zero-width window (`before=0,
+            after=0`).")
+    }
+    after <- 0L
+  }
+
+  # Make a complete date sequence between min(x$time_value) and max(x$time_value).
+  date_seq_list <- full_date_seq(x, before, after, time_step)
+  all_dates <- date_seq_list$all_dates
+  pad_early_dates <- date_seq_list$pad_early_dates
+  pad_late_dates <- date_seq_list$pad_late_dates
+
+  # `frollmean` is 1-indexed, so create a new window width based on our
+  # `before` and `after` params.
+  m <- before + after + 1L
+
+  if (is.null(names_sep) && !as_list_col) {
+    if (length(new_col_name) != length(col_name)) {
+     cli_abort(
+        "`new_col_name` must be the same length as `col_name` when `names_sep` is NULL.",
+        class = "epiprocess__epi_slide_mean__col_name_length_mismatch",
+        epiprocess__new_col_name = new_col_name,
+        epiprocess__col_name = col_name
+      )
+    }
+    result_col_name <- new_col_name
+  } else {
+    if (length(new_col_name) != 1L && length(new_col_name) != length(col_name)) {
+     cli_abort(
+        "`new_col_name` must be either length 1 or the same length as `col_name`.",
+        class = "epiprocess__epi_slide_mean__col_name_length_mismatch_and_not_one",
+        epiprocess__new_col_name = new_col_name,
+        epiprocess__col_name = col_name
+      )
+    }
+    result_col_name <- paste(new_col_name, col_name, sep = names_sep)
```

Comment:
Removed `as_list_col` option. I don't think the list-col behavior is useful and it doesn't do the same thing as the `epi_slide` version.

Reply:
For the first issue,

```
jhu_csse_daily_subset %>%
  group_by(geo_value) %>%
  epi_slide_mean(
    c("cases", "death_rate_7d_av"),
    new_col_names = "my_slide_value",
    names_sep = NULL,
    before = 6
  )
```

generates the error "! `new_col_names` must be the same length as `col_names` when `names_sep` is NULL" on purpose (hopefully that's the error you were seeing).

If `names_sep = NULL`, output column names are just set to `new_col_names`, so `new_col_names` must be the same length as requested `col_names` to avoid duplicate output column names. If `names_sep = <non-NULL>`, output column names are `new_col_names + <separator> + col_names`.

I'll make the error message more actionable, but let me know if you were having a different problem.